frontend/src/metabase/core/components/TabContent/TabContent.tsx

```diff
@@ -0,0 +1,28 @@
+import React, { ReactNode, useMemo } from "react";
+import { useUniqueId } from "metabase/hooks/use-unique-id";
+import { TabContext, TabContextType } from "../Tab";
+
+export interface TabContentProps<T> {
+  value?: T;
+  children?: ReactNode;
+  onChange?: (value: T) => void;
+}
+
+const TabContent = function TabContent<T>({
+  value,
+  children,
+  onChange,
+}: TabContentProps<T>) {
+  const idPrefix = useUniqueId();
+  const context = useMemo(() => {
+    return { value, idPrefix, onChange };
+  }, [value, idPrefix, onChange]);
+
+  return (
+    <TabContext.Provider value={context as TabContextType}>
```

Comment:
Should we maybe set `TabContextType` on `useMemo` instead of doing casting? Like:

```js
const context = useMemo<TabContextType>(...)
```

Reply:
Unfortunately, the casting is required because the type is not the same - `TabContextType<T>` vs `TabContextType<undefined`. Since `Provider` cannot use generic types, this is the only workaround I know.